### PR TITLE
frontend  - redux-saga implementation + unit testing

### PR DESCRIPTION
**Pull request by ptitdam2001**
- implementation of redux-saga
- usage of redux-saga for series list
- Integration of unit testing in frontend

### CODE REVIEW

File: client/web/src/features/series/serieSaga.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    type: actionTypes.SET_ALL_SERIES,
+    series,
+  })
+}
+
+function* actionWatcher() {
+  yield takeLatest(actionTypes.FETCH_ALL_SERIES, fetchSeries)
+}
+
+export default function* userSaga() {
```
Comment (bat3) [inline client/web/src/features/series/serieSaga.ts]: should be seriesSaga

File: client/web/src/features/series/serieSaga.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { put, takeLatest, call, all } from 'redux-saga/effects'
+
+import * as actionTypes from './actionTypes'
+import SerieService from './serieService'
+
+function* fetchSeries() {
+  const { data: series } = yield call([SerieService, 'getAll'])
```
Comment (bat3) [inline client/web/src/features/series/serieSaga.ts]:
```suggestion
  const { data: series } = yield call(SerieService.getAll()])
```
As in userSaga

File: client/web/src/features/series/actionTypes.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+export const SET_CURRENT_SERIE = 'SET_CURRENT_SERIE'
```
Comment (bat3) [inline client/web/src/features/series/actionTypes.ts]:
```suggestion
export const SET_CURRENT_SERIE = 'SET_CURRENT_SERIES'
```
'Serie' doesn't exist in English

File: client/web/src/features/series/serieReducer.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import {
+  UNSET_CURRENT_SERIE,
+  SET_CURRENT_SERIE,
+  SET_ALL_SERIES,
+} from './actionTypes'
+import { SerieActionTypes, SerieState } from './types'
+
+const initialState = {
+  list: [],
+  serie: null,
```
Comment (bat3) [inline client/web/src/features/series/serieReducer.ts]: 'Serie' doesn't exist in English

File: client/web/src/models/serie.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+export interface ISerie {
```
Comment (bat3) [inline client/web/src/models/serie.ts]: 'Serie' doesn't exist in English

Need to check all files ;-)

File: client/web/src/features/user/userService.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import { IUser } from '../../models/user'
```
Comment (bat3) [inline client/web/src/features/user/userService.ts:1]: You don't think we can create a crudService by setting only the type (Serie or User) and the base URL ?
It seems to be same code between userService and seriesService.

**Review by bat3 — CHANGES_REQUESTED** [review]

File: client/web/src/features/series/serieSaga.ts (hunk 2 above)
Comment (ptitdam2001) [reply client/web/src/features/series/serieSaga.ts]: it is not the good usage, it does not work

File: client/web/src/features/user/userService.ts (hunk 6 above)
Comment (ptitdam2001) [reply client/web/src/features/user/userService.ts:1]: I don't think because we will have specific methods (for example login method)

Comment (ptitdam2001) [reply client/web/src/features/user/userService.ts:1]: it is that why I implemented one interface ICrudService

**Comment by ptitdam2001** [comment]
I finished the development for that branch, I am waiting for yours feedbacks. It may be complicated for non front developers.

To summary, I implemented some unit tests with Jest and we cover all about series . The global coverage is 50% for the moment. Morever, I implemented redux-saga which it is a sort of event listener inside our front store. It manages side-effects : asynchronous things like data fetching and impure things like accessing the browser cache

File: client/web/src/features/series/seriesSaga.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    type: actionTypes.SET_ALL_SERIES,
+    series,
+  })
+}
+
+export function* actionWatcher() {
+  yield takeLatest(actionTypes.FETCH_ALL_SERIES, fetchSeries)
+}
+
+export default function* serieSaga() {
```
Comment (bat3) [inline client/web/src/features/series/seriesSaga.ts]:
```suggestion
export default function* seriesSaga() {
```

File: client/web/src/features/series/seriesService.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import { ISeries } from '../../models/series'
+
+import { ICrudService } from '../common/ICrudService'
+import { HttpRequest } from '../common/HttpRequest'
+
+export class SerieService implements ICrudService<ISeries> {
```
Comment (bat3) [inline client/web/src/features/series/seriesService.ts]:
```suggestion
export class SeriesService implements ICrudService<ISeries> {
```

File: client/web/src/features/user/userSaga.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import { put, takeLatest, call, all } from 'redux-saga/effects'
+
+import * as actionTypes from './actionTypes'
+import UserService from './userService'
+
+function* fetchUsers() {
+  const users = yield call(UserService.getAll)
```
Comment (bat3) [inline client/web/src/features/user/userSaga.ts:7]: Not consitente with seriesSaga:
`const { data: series } = yield call([serieService, 'getAll'])`

File: client/web/src/features/series/seriesService.spec.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
+import { ISeries } from '../../models/series'
+import serieService, { SerieService } from './seriesService'
+import { HttpRequest } from '../common/HttpRequest'
+
+jest.mock('../common/HttpRequest')
+
+describe('SerieService', () => {
```
Comment (bat3) [inline client/web/src/features/series/seriesService.spec.ts]:
```suggestion
describe('SeriesService', () => {
```

File: client/web/src/i18n.ts [hunk 11]
```diff
@@ -25,6 +25,7 @@ i18n
     interpolation: {
       escapeValue: false, // react already safes from xss
     },
+    // initImmediate: false,
```
Comment (bat3) [inline client/web/src/i18n.ts]: To delete

File: client/web/src/features/user/userSaga.ts (hunk 9 above)
Comment (ptitdam2001) [reply client/web/src/features/user/userSaga.ts:7]: User feature is not up to date, I will create a new branch about user is starting

**Review by bat3 — APPROVED** [review]

**Pull request merged**